src/Microsoft.Kusto.ServiceLayer/DataSource/KustoDataSource.cs

```diff
@@ -626,93 +612,311 @@ internal IEnumerable<ColumnInfo> GetColumnMetadata(string databaseName)
+        {
+            foreach (var columnGroup in columnsGroupByTable)
+            {
+                string tableName = columnGroup.Key;
+                var columns = new List<ColumnMetadata>();
+                foreach (ColumnInfo columnInfo in columnGroup)
+                {
+                    var column = new ColumnMetadata
+                    {
+                        ClusterName = this.ClusterName,
+                        DatabaseName = databaseName,
+                        TableName = tableName,
+                        MetadataType = DataSourceMetadataType.Column,
+                        MetadataTypeName = DataSourceMetadataType.Column.ToString(),
+                        Name = columnInfo.Name,
+                        PrettyName = columnInfo.Name,
+                        Urn = $"{this.ClusterName}.{databaseName}.{tableName}.{columnInfo.Name}",
+                        DataType = columnInfo.DataType
+                    };
+                    
+                    columns.Add(column);
+                }
+
+                _columnMetadata[GenerateMetadataKey(databaseName, tableName)] = columns;
+            }
+        }
+
+        private void SetTableMetadata(string databaseName, IEnumerable<ColumnInfo> columnInfos, string tableFolderKey)
+        {
+            var columnInfoTables =
+                columnInfos.Where(x => !string.IsNullOrWhiteSpace(x.Table)
+                                       && string.IsNullOrWhiteSpace(x.Name));
+
+            var tables = new Dictionary<string, List<TableMetadata>>();
+            foreach (var columnInfoTable in columnInfoTables)
+            {
+                var stringBuilder = new StringBuilder(tableFolderKey);
+
+                if (!string.IsNullOrWhiteSpace(columnInfoTable.Folder))
+                {
+                    stringBuilder.Append($".{columnInfoTable.Folder}");
+                }
+
+                var tableKey = $"{stringBuilder}";
+
+                var tableMetadata = new TableMetadata
+                {
+                    ClusterName = ClusterName,
+                    DatabaseName = databaseName,
+                    MetadataType = DataSourceMetadataType.Table,
+                    MetadataTypeName = DataSourceMetadataType.Table.ToString(),
+                    Name = columnInfoTable.Table,
+                    PrettyName = columnInfoTable.Table,
+                    Folder = columnInfoTable.Folder,
+                    Urn = $"{stringBuilder}.{columnInfoTable.Table}"
+                };
+
+                if (tables.ContainsKey(tableKey))
+                {
+                    tables[tableKey].Add(tableMetadata);
+                }
+                else
+                {
+                    tables[tableKey] = new List<TableMetadata>{tableMetadata};
+                }
+            }
+            
+            foreach (var table in tables)
+            {
+                _tableMetadata[table.Key] = table.Value;    
```

Comment:
Shouldn't we append tables if the value already exists? If you do not expect value to exist, please add an assert here.